service-base/src/main/java/org/eclipse/hono/service/resourcelimits/PrometheusBasedResourceLimitChecks.java

```diff
@@ -467,8 +467,11 @@ private void checkMessageLimit(
                         .map(cachedValue -> Future.succeededFuture((long) cachedValue))
                         .orElseGet(() -> executeQuery(queryParams, span).map(bytesConsumed -> addToCache(limitsCache, key, bytesConsumed)))
                         .map(bytesConsumed -> {
+                            final long totalBytesConsumed = bytesConsumed + payloadSize;
+                            limitsCache.put(key, bytesConsumed + payloadSize);
```

Comment:
shouldn't this only be put to the cache if the limit is not exceeded?

Reply:
I thought about this too. But since I observed the Prometheus metrics are also increasing even in case the message limit was hit, I though it is more consistent this way.

Reply:
If I am not mistaken then the metrics are not increased in case of a server side error, e.g. registry not available. We only count the message bytes in if we are able to process the message or if the reason for a failure to process lies with the client (indicated by a ClientErrorException). Maybe we can follow the same logic here?

Reply:
As I said, I observed this locally on my machine. 

Let me double-check, maybe there was something else going on which lead to the increasing byte count in Prometheus.

Reply:
I still do observe this on my machine. Here's what I did:

1. Started a local Hono instance by using the packages project
1. Created a new tenant and device
1. Configured a message limit (1024) for the created tenant
1. Send messages via HTTP adapter until the message limit was hit
1. Waited until the cache timed out (set locally to 15 seconds)
1. Send one more message
1. Observed that the `bytesConsumed` increased in `PrometheusBasedResourceLimitChecks`

Looking at the code I think this works as programmed, since `metrics.reportTelemetry(..)` is called in `.recover(...)` meaning also in an error case (see https://github.com/eclipse/hono/blob/master/adapters/http-vertx-base/src/main/java/org/eclipse/hono/adapter/http/AbstractVertxBasedHttpProtocolAdapter.java#L794).

Also the values queried from Prometheus do include `FORWARDED` and `UNPROCESSABLE` (according my understanding this value is set in case of the returned 429) - hence the bytes reported for a message above the limit will add up.

Hopefully I did not miss something obvious. If you can confirm my observations and findings in the code I suggest to create a bug and only update the cache if the message limit is not hit.

Reply:
It makes sense to report every request using *Metrics.reportTelemetry* in order to collect data on all requests. Another question is, which payload the customer needs to pay for. The overall number of bytes uploaded by a customer may indeed by larger than the number determined by the ResourceLimitsChecker, simply because we do not charge the customer for data that has not been processed due to a server error. That is why the query only includes `FORWARDED` and `UNPROCESSABLE` (client error).
This means that the data volume check should not add data to its local counter that cannot be processed due to a server error. Otherwise the calculation of the limit would be inconsistent with when we use consumption data retrieved freshly from Prometheus, right?

Reply:
Yes, that is exactly my point: this would be inconsistent.

I fully agree on updating the cache only when the limit is hit.

But, if am not completely off track 😄 it's just that the Prometheus-based limit calculation does not work like this:

In case the message limit is hit, actually a `ClientErrorException(HttpResponseStatus.TOO_MANY_REQUESTS.code()))` is thrown (see: https://github.com/eclipse/hono/blob/master/service-base/src/main/java/org/eclipse/hono/service/AbstractProtocolAdapterBase.java#L662), which is then "catched" in the `.recover(...)` callback I mentioned in my previous comment, consequently being reported as a metric with a `UNPROCESSABLE` tag.

When the next message comes in Prometheus is queried by both tags `FORWARDED` and `UNPROCESSABLE`. Which makes me think that the actually rejected message adds to the payload - since we put the label "client error" on it.

Does that make any sense? 😅

Reply:
> When the next message comes in Prometheus is queried by both tags FORWARDED and UNPROCESSABLE. Which makes me think that the actually rejected message adds to the payload - since we put the label "client error" on it.

Yes, that is expected and intended behavior. We notify the device/client about the fact that the data volume limit has been exceeded. However, we cannot prevent the device/client from sending additional messages with more payload. Once we perform the check if the limit is reached/exceeded, the data has already been transferred to our servers and thus produced costs. The only thing we can do is to charge the customer for this erratic behavior.

> I fully agree on updating the cache only when the limit is hit.

This is not the problem. The problem is that when an adapter invokes the *checkMessageLimit* method, we do not know if the data will need to be accounted for (FORWARDE or UNPROCESSABLE) or not (UNDELIVERABLE). In order to be consistent with the *current bytes consumed* as retrieved form Prometheus, we should only add the message payload length to the local counter in the former case (FORWARDED or UNPROCESSABLE).

Reply:
> Yes, that is expected and intended behavior. 

Okay, that was what I was confused about.

I'm clear about the rest now. Thank you for the explanation.

Reply:
As discussed I reworked my changes so that the update of the cache and the check for the limit is separated as it is done in Hono's code up to now. By this it is also assured that the byte counter in the cache is only increased for FORWARDED and UNPROCESSABLE outcomes.

Reply:
I just took a deeper look on the changes done with https://github.com/eclipse/hono/commit/385c5a45a15a09ceacc675d892d3bf86d4a40941.

I will have to do a rework of my rework before being able to rebase - so reviewing my latest commit is rather pointless until this is done.

Reply:
Done! Review can continue 😃